# CV-TMLE: cross-fitting and navigator teams

This notebook fits the same average treatment effect as the
[point-treatment tutorial](point-treatment-tmle.ipynb), with flexible learners. Each step shows its
code, its output, and what the output tells you.
[CV-TMLE and cross-fitting](../technical-reference/cv-tmle.md) gives the constructions and their
fold arithmetic.

## The applied question

The point-treatment tutorial fit boosted learners and reported one interval. A reviewer of the
network report asks two questions. The learners predict held-out patients well, so why split the
sample? Patients come from two hundred navigator teams, so are they independent observations?

The question itself does not change. The program sponsor still asks how much the mean 30-day
transition score would change if every eligible discharge received the offer.

## What you will learn

| after this notebook you can | the step that shows it |
| --- | --- |
| state which condition cross-fitting addresses, and which conditions it does not | why this method |
| fit CV-TMLE with an explicit fold count, and read the construction it reports | Step 5 |
| compare in-sample and cross-fitted intervals, and read how often each one covers | Step 6 |
| declare navigator teams, and read what changes in the interval and the folds | Steps 7 and 8 |
| reuse one outer split to compare fits on identical folds | Step 9 |
| tell the three CV-TMLE constructions apart | Step 10 |
| read the diagnostics and the spread across fold draws | Steps 11 and 12 |
| read an omitted-confounding analysis for the cross-fitted estimate, and the estimator it rests on | Step 13 |

## Why this method

| your situation | what this method buys | what it costs |
| --- | --- | --- |
| a flexible learner for either nuisance | an interval without a Donsker restriction, under the [remaining conditions](../technical-reference/cv-tmle.md#what-this-solves) | one nuisance fit per outer fold, ten by default |
| patients nested in navigator teams | teams stay intact in every split, and the variance uses team totals | a wider interval, because teams, not patients, are the independent units. With fewer teams than folds, the fold count drops |

Classical interval proofs that reuse observations control an empirical-process term with a
complexity condition such as a Donsker condition. Rich, tuned learners need not satisfy that
condition, even when they predict well. Cross-fitting breaks this data-reuse path because each
patient's nuisance prediction comes from a model that never saw that patient. It addresses data
reuse only. The [CV-TMLE reference](../technical-reference/cv-tmle.md#what-this-solves) lists the
conditions that your learners and your study design must supply.

| term | plain meaning |
| --- | --- |
| estimand | the number the question asks for, written before any model is chosen. [Estimands](../user-guide/estimands.md) defines the available questions |
| nuisance | a model the estimate needs but the question does not ask about. Here, the outcome regression Q and the treatment mechanism g. [Point-treatment TMLE](../technical-reference/point-treatment-tmle.md) defines both |
| cross-fitting | each row's nuisance prediction comes from models fit without that row. [CV-TMLE](../technical-reference/cv-tmle.md) defines it |
| targeting | a small update to the outcome model, weighted by the treatment model, that removes first-order bias. [Methods and learners](../user-guide/methods-learners.md#targeting-and-bounds) defines the update |
| outer fold | one part of the sample. The models for its rows train on the other parts |
| influence curve | how much each row moves the estimate. Its variance gives the standard error. [Inference](../technical-reference/inference.md) defines it |
| cluster | a group of rows that are not independent of each other. Here, one navigator team |
| split plan | the realized fold label of every row, which a later fit can reuse |
| calibration slope | the slope of a logistic recalibration of the out-of-fold predicted probabilities, with one intercept per fold. A slope of 1 means that the spread of the predictions matches the observed rates. A slope below 1 means the predictions are too extreme |
| effective sample size | how many equally weighted rows the weights are worth. A ratio of 25% means the weights are as uneven as if a quarter of the rows counted |

## Step 1: set up

The setup imports the learners and prints the installed `cleverly` version. Every fit below passes
its learners, fold count, and random seed explicitly, so a rerun reproduces the stored outputs.

In [1]:
import numpy as np
import pandas as pd
from sklearn.ensemble import HistGradientBoostingClassifier, HistGradientBoostingRegressor
from sklearn.linear_model import LogisticRegression

import cleverly

pd.set_option("display.width", 130)
pd.set_option("display.max_columns", 20)
print("cleverly", cleverly.__version__)

cleverly 0.1.2


**What this output tells you.** The stored outputs in this notebook came from the version named
above. A different version can print different numbers.

## Step 2: the data

The law is the point-treatment tutorial's law, drawn with a new seed. Its nuisance functions are
nonlinear, which is the case where a flexible learner earns its cost. `navigation_data` draws the
rows under the program's column names. The code prints the column names, the first rows, and the
true ATE.

In [2]:
from cleverly.datasets import navigation_data

frame, truth = navigation_data(n=3_000, seed=34)
print("rows and columns:", frame.shape)
print(list(frame.columns))
print(frame.head().round(3))
print()
print(f"population ATE: {truth['ate']:.3f}")

rows and columns: (3000, 6)
['transition_score', 'transition_navigation', 'discharge_risk', 'prior_utilization', 'medication_burden', 'age']
   transition_score  transition_navigation  discharge_risk  prior_utilization  medication_burden    age
0             0.454                    0.0          -0.040             -1.258              2.574  0.482
1             0.791                    1.0           0.644             -0.208              0.058  0.337
2             0.735                    1.0           0.195             -0.609              0.615  0.490
3             0.295                    0.0          -0.118             -0.719             -0.879 -1.008
4             0.642                    1.0           0.277             -0.351              0.871 -0.888

population ATE: 0.163


**What this output tells you.** Each row is one discharge. The helper `navigation_data` draws the
rows from the generator `make_nonlinear_bounded` and gives the columns the program's names.
`transition_navigation` is 1 for an offer and 0 for usual support. The four baseline covariates
are independent standard-normal draws. The score is a share of the maximum score.

| feature of the law | what it means for this page |
| --- | --- |
| both nuisance functions are nonlinear | a flexible learner can earn its cost here, which is the case cross-fitting exists for |
| the true propensity lies between 5% and 95% | every patient has at least a 5% chance of each arm. This is strong positivity, so the ATE has a finite efficiency bound |
| the score lies between 0 and 1 | the support belongs to the instrument, and the protocol records the scale. Step 5 declares it to the fit |

The true ATE is 0.163. A real program has no `truth`. Every comparison against it below is a
teaching device.

## Step 3: write the protocol

This page keeps the [shared study design](index.md#the-shared-study-design) and changes only the
estimator. Cross-fitting is an estimation choice, and it must not touch the question. The code
therefore uses `navigation_protocol()` unchanged, as the point-treatment tutorial does.

In [3]:
from cleverly.datasets import navigation_protocol

protocol = navigation_protocol()
print("\n".join(protocol.summary_lines()))

causal study protocol: schema 1; 623fc2c240615d26
target population: Adults with a discharge-home order at a participating hospital during the enrollment period
eligibility: ['Age 18 years or older', 'Discharge home ordered at a participating hospital']
time zero: Discharge-home order, after baseline measurement and before the navigation offer
treatment strategies: ['Offer standard transition navigation', 'Provide usual discharge support']
treatment versions: ['Bedside transition plan and two scheduled navigator contacts within 30 days', 'No access to the transition-navigation offer']
outcome: Patient-reported transition score, as a share of the maximum score
horizon: 30 days after discharge
intercurrent-event handling: ['Use the transition score regardless of readmission', 'Analyze the offer regardless of completed contacts', 'The protocol scores death before day 30 as the worst transition score (composite strategy)']
interference unit: Individual patient
assumption rationale: ['The r

**What this output tells you.** The first line gives the fingerprint `623fc2c240615d26`. The
point-treatment tutorial prints the same fingerprint, because the two pages record the same design.
[Point-treatment TMLE](point-treatment-tmle.ipynb) reads each field.

The outcome field names the score's scale. Step 5 reads the declared support from that field
rather than from the data.

The interference unit stays the individual patient. Shared navigator teams are a statistical
dependence, not interference. `StudyProtocol` has no field for that dependence. Step 7 declares it
through the design's `cluster=` role instead.

Clustering is not an interference adjustment. If patients compete for navigator slots, potential
outcomes can depend on other assignments. A cluster-robust standard error cannot repair that
causal-design failure.


## Step 4: design and identification

The design names the column roles, and the estimand names the contrast. Identification returns the
observed-data formula, the nuisances it needs, and the assumptions that make it causal.

In [4]:
from cleverly import ATE, CausalStudy, PointTreatment

study = CausalStudy(
    frame,
    design=PointTreatment(
        outcome="transition_score",
        treatment="transition_navigation",
        adjustment=("discharge_risk", "prior_utilization", "medication_burden", "age"),
    ),
    protocol=protocol,
)
effect = study.identify(ATE(reference=0))
print(effect.summary(protocol="fingerprint"))

average treatment effect, E[Y^a] - E[Y^reference]
identified by explicit-adjustment: E_W[E(transition_score | transition_navigation=a, W)] - E_W[E(transition_score | transition_navigation=0, W)] for a in [1]
adjustment/history: ['discharge_risk', 'prior_utilization', 'medication_burden', 'age']
required nuisances: ['outcome_regression', 'treatment_mechanism']
assumptions:
  - consistency: Y = Y^a when A = a
  - no interference: one unit's potential outcome does not depend on other units' treatment assignments
  - no unmeasured confounding: Y^a is independent of A given W
  - positivity: P(transition_navigation = a | W) > 0 almost surely for every supported treatment level a in [0, 1]
causal study protocol: schema 1; 623fc2c240615d26


**What this output tells you.** The estimand, the formula, the two required nuisances, and the four
assumptions match the point-treatment tutorial. The summary also prints the fingerprint of the
stored protocol.

Nothing in this output names folds or learners. The fold choice belongs to the method in the next
step, so a change of estimator cannot change the question.

## Step 5: estimate with cross-fitting

Cross-fitting is configured as a named group. The configuration is written out in full.

| setting | value | what it does |
| --- | --- | --- |
| `outcome_learner` | gradient boosting | fits Q, the expected score given the offer and the covariates |
| `treatment_learner` | a shallow, regularized booster | fits g, the probability of an offer given the covariates |
| `CrossFitting(n_folds=5)` | five outer folds | predicts each row from models fit on the other four folds |
| `Targeting(q_bounds=(0.0, 1.0))` | the declared score support | fixes the outcome scale before any fold is drawn |
| `Runtime(random_state=34, n_jobs=1)` | fixed seed, one process | makes the fold draw and the fit reproducible |

The package default is ten outer folds. Five keeps this notebook fast.

The outcome learner is the default gradient-boosting regressor. It is the kind of flexible learner
that this page is about. The treatment learner uses trees of depth 2, a small learning rate, and an
L2 penalty.

Do not use the default `HistGradientBoostingClassifier` as a treatment learner below
10,000 rows. Below that size, scikit-learn does not stop it early, and its fitted propensities are
too extreme. The [rules for code cells](../development/example-notebooks.md#rules-for-code-cells)
state this requirement. Step 11 checks the calibration of the fitted propensities.

Declare the known outcome support as `q_bounds` whenever you cross-fit a continuous outcome. A
cross-fitted fit of a continuous outcome needs it. The declared support is part of the
cross-fitting rule, not a tuning choice. It sets the outcome scale before any fold is drawn.
Otherwise, each fold's outcome scale would come partly from the rows that the fold holds out.
The name `declared_support` carries the same object to every fit below, so no comparison mixes two
scales.

The boosted learners have no inner learner folds, so `learner_folds` does not apply. The
[methods guide](../user-guide/methods-learners.md#two-fold-layers) explains the two fold layers.

The summary prints the protocol fingerprint. Read the construction line, the footer, and the
estimate table.

In [5]:
from cleverly import CrossFitting, ModelSpec, Runtime, Targeting, TMLEMethod


def shallow_booster():
    """A shallow, regularized boosted propensity learner."""
    return HistGradientBoostingClassifier(
        max_depth=2,
        learning_rate=0.05,
        max_iter=200,
        l2_regularization=1.0,
        random_state=34,
    )


boosted = ModelSpec(
    outcome_learner=HistGradientBoostingRegressor(random_state=34),
    treatment_learner=shallow_booster(),
)
declared_support = Targeting(q_bounds=(0.0, 1.0))
cross_fitted = effect.estimate(
    method=TMLEMethod(
        models=boosted,
        cross_fitting=CrossFitting(n_folds=5),
        targeting=declared_support,
        runtime=Runtime(random_state=34, n_jobs=1),
    )
)
print(cross_fitted.summary(protocol="fingerprint"))
print()
print(f"population ATE: {truth['ate']:.3f}")

Targeted maximum likelihood estimation
n = 3000; covariates = 4; P(A=1) = 0.4597
causal estimand: average treatment effect, E[Y^a] - E[Y^reference]
identification: explicit-adjustment; E_W[E(transition_score | transition_navigation=a, W)] - E_W[E(transition_score | transition_navigation=0, W)] for a in [1]
required nuisances: outcome_regression, treatment_mechanism
identification assumptions: consistency: Y = Y^a when A = a; no interference: one unit's potential outcome does not depend on other units' treatment assignments; no unmeasured confounding: Y^a is independent of A given W; positivity: P(transition_navigation = a | W) > 0 almost surely for every supported treatment level a in [0, 1]
causal study protocol: schema 1; 623fc2c240615d26
outcome family: gaussian; fluctuation: logistic (iterative)
stacked CV-TMLE (Levy): nuisances cross-fitted over 5 folds; targeting: pooled
propensity truncated to [0.0114, 0.9886]; ATT/ATC to [0.025, 0.975]
data 32c6f838ef95b45f | folds c7f4eafaaf31

**What this output tells you.** The construction line reads
`stacked CV-TMLE (Levy): nuisances cross-fitted over 5 folds; targeting: pooled`. It names the
construction, the number of outer folds, and the targeting scheme. Step 10 compares the other
constructions.

The summary prints no `outcome scaled` line. The declared support already matches the score's own
scale, so the fit rescales nothing.

The estimate is 0.15731 with a standard error of 0.005444. The 95% interval is
[0.14664, 0.16798], and it contains the true ATE of 0.163. That is one draw, not a coverage result.

A review probe, `reviews/notebook-review/probes/cross-fitting-final/sweep.py`, repeats this fit and the Step 6 fit on 200 draws of the same
law. The cross-fitted interval contained the true ATE on 188 of them. Its mean standard error was
0.93 times the standard deviation of the 200 estimates. The probe checks this page's
configuration. It is not a registered study.

The footer line records the data fingerprint `32c6f838ef95b45f` and the fold fingerprint
`c7f4eafaaf319249`. Step 9 uses both.

## Step 6: the failure mode, an interval that is too narrow

Now fit the same learners with the splitting turned off. `CrossFitting(enabled=False)` is not a
tuning knob. It selects ordinary TMLE rather than CV-TMLE. The fit keeps the same declared
support, so cross-fitting is the only difference between the two. The code prints both fits side
by side, and it marks whether each interval contains the true ATE.


In [6]:
in_sample = effect.estimate(
    method=TMLEMethod(
        models=boosted,
        cross_fitting=CrossFitting(enabled=False),
        targeting=declared_support,
        runtime=Runtime(random_state=34, n_jobs=1),
    )
)


def summary_line(result, marker):
    """The first line of the result summary that contains ``marker``."""
    return next(line.strip() for line in result.summary().splitlines() if marker in line)


def side_by_side(fits, target=None, columns=("psi", "std_err", "ci_lower", "ci_upper")):
    """One row per labelled fit, and whether each interval contains ``target`` when given."""
    table = pd.concat({label: fitted.to_frame() for label, fitted in fits.items()})
    table = table.droplevel(1)[list(columns)]
    if target is not None:
        table["covers truth"] = table["ci_lower"].le(target) & table["ci_upper"].ge(target)
    return table


print(summary_line(in_sample, ": nuisances "))
comparison = side_by_side(
    {"no cross-fitting": in_sample, "cross-fitted": cross_fitted}, truth["ate"]
)
print(comparison.round(4))
se_ratio = in_sample["ate"].std_error / cross_fitted["ate"].std_error
print(f"in-sample SE / cross-fitted SE: {se_ratio:.2f}")
for label, fitted in (("in-sample", in_sample), ("cross-fitted", cross_fitted)):
    miss = abs(truth["ate"] - fitted["ate"].psi) / fitted["ate"].std_error
    print(f"{label} distance from the truth, in its standard errors: {miss:.1f}")
print(f"population ATE: {truth['ate']:.3f}")

TMLE (in-sample nuisances): nuisances fitted in-sample (cross_fit=False)
                     psi  std_err  ci_lower  ci_upper  covers truth
no cross-fitting  0.1553   0.0040    0.1476    0.1631          True
cross-fitted      0.1573   0.0054    0.1466    0.1680          True
in-sample SE / cross-fitted SE: 0.73
in-sample distance from the truth, in its standard errors: 1.9
cross-fitted distance from the truth, in its standard errors: 1.0
population ATE: 0.163


**What this output tells you.** The first line confirms that the new fit used in-sample nuisances.
The table then compares the two fits on this draw.

| quantity | no cross-fitting | cross-fitted |
| --- | --- | --- |
| point estimate | 0.1553 | 0.1573 |
| standard error | 0.0040 | 0.0054 |
| 95% interval | (0.1476, 0.1631) | (0.1466, 0.1680) |
| distance from the true ATE of 0.163, in the fit's own standard errors | 1.9 | 1.0 |
| contains the true ATE | yes | yes |

On this draw, the in-sample standard error is 0.73 times the cross-fitted one. Both intervals
contain the true ATE. The truth lies 1.9 in-sample standard errors from the in-sample estimate, so
that interval nearly misses.

One draw cannot show a coverage rate. The review probe `reviews/notebook-review/probes/cross-fitting-final/sweep.py` repeats both fits on
200 draws of the same law.

| over 200 draws | no cross-fitting | cross-fitted |
| --- | --- | --- |
| intervals that contained the true ATE | 153 | 188 |
| mean standard error divided by the standard deviation of the estimates | 0.70 | 0.93 |
| mean estimate minus the true ATE | -0.0024 | 0.0003 |

The in-sample interval missed on about one draw in four. That is the failure mode itself, not
sampling variation. A calibrated 95% interval misses on about one draw in 20. The probe shows two
causes. The in-sample standard error is too small, and the in-sample estimate sits slightly low.
On 90% of the draws, the in-sample standard error was between 0.68 and 0.73 times the cross-fitted
one.

The in-sample fit also gives more even observed-arm weights, which Step 11 prints. These values do
not isolate the empirical-process term. The
[CV-TMLE reference](../technical-reference/cv-tmle.md#what-this-solves) explains the data-reuse
condition.

The registered
[stacked CV-TMLE study](../technical-reference/method-evidence/stacked-point-treatment-cv-tmle.md#theory-properties)
fits this same law 400 times at n = 500. It uses a fully grown regression tree for Q, a logistic
model for g, ten folds, and the same declared support. The in-sample control covers 0.5225 of the
time, and the cross-fitted fit covers 0.945. The study's
[property results](https://github.com/esbraun/cleverly-tmle/blob/main/tests/canonical/tmle3_cvtmle/properties.csv)
record the design.

## Step 7: the second failure mode, patients are not independent

The program has two hundred navigator teams with fifteen patients each. Teams differ in ways the
recorded covariates do not capture, such as supervisor practice and local follow-up quality. A
shared team difference in the benefit of navigation makes the influence values of one team move
together.

The `cluster=` role declares the teams. The team data come from `make_clustered`, which is a
separate synthetic law with its own covariates and its own truth. The code gives its two covariates
program names, but they are not the Step 2 covariates. The outcome is the top-box indicator, which
is 1 when a patient reports the top box on the transition score.

The outcome is binary, so its support is known, and the team fits cross-fit without a declared
support. Step 8 then reads the realized split. The protocol changes only the outcome field. The
code prints the new fingerprint and that field.

The team fit uses logistic learners for both nuisances. The true treatment mechanism of this law is
a main-terms logistic model in the two covariates, so the treatment learner has the right form.

The shared team factor shifts both the baseline score and the benefit, and it is strong. In this
law, about 46% of teams have a negative mean effect of the offer (review probe
`reviews/notebook-review/probes/cross-fitting-final/truth.py`). The pooled ATE therefore hides large team differences.

In this law the team factor does not change who receives an offer. A team factor that also drives
offers is an unmeasured confounder, and no cluster declaration repairs it.

The code builds one design without the cluster role and a copy with `cluster="navigator_team"`.
Nothing else differs between the two fits.

In [7]:
from dataclasses import replace

from cleverly.datasets import make_clustered

team_frame, team_truth = make_clustered(n=3_000, seed=34, cluster_size=15, family="binomial")
team_frame = team_frame.rename(
    columns={
        "Y": "transition_top_box",
        "A": "transition_navigation",
        "W1": "discharge_risk",
        "W2": "medication_burden",
        "cluster": "navigator_team",
    }
)
team_protocol = replace(protocol, outcome="Top-box patient-reported transition score")
simple = TMLEMethod(
    models=ModelSpec(
        outcome_learner=LogisticRegression(max_iter=1000, random_state=34),
        treatment_learner=LogisticRegression(max_iter=1000, random_state=34),
    ),
    cross_fitting=CrossFitting(n_folds=5),
    runtime=Runtime(random_state=34, n_jobs=1),
)
patients_design = PointTreatment(
    outcome="transition_top_box",
    treatment="transition_navigation",
    adjustment=("discharge_risk", "medication_burden"),
)
teams_design = replace(patients_design, cluster="navigator_team")

ignoring = (
    CausalStudy(team_frame, design=patients_design, protocol=team_protocol)
    .identify(ATE(reference=0))
    .estimate(method=simple)
)
clustered = (
    CausalStudy(team_frame, design=teams_design, protocol=team_protocol)
    .identify(ATE(reference=0))
    .estimate(method=simple)
)

team_comparison = side_by_side(
    {"patients as units": ignoring, "teams as units": clustered}, team_truth["ate"]
)
print(team_comparison.round(4))
team_se_ratio = clustered["ate"].std_error / ignoring["ate"].std_error
print(f"teams SE / patients SE: {team_se_ratio:.2f}")
print("navigator teams:", clustered.data.n_clusters)
print(f"population ATE: {team_truth['ate']:.3f}")
print("team protocol fingerprint:", team_protocol.fingerprint)
print("team protocol outcome:", team_protocol.outcome)

                      psi  std_err  ci_lower  ci_upper  covers truth
patients as units  0.1237   0.0195    0.0856    0.1619          True
teams as units     0.1256   0.0301    0.0667    0.1845          True
teams SE / patients SE: 1.54
navigator teams: 200
population ATE: 0.104
team protocol fingerprint: 7df34d840b4e8d1a
team protocol outcome: Top-box patient-reported transition score


**What this output tells you.** The table compares one law fitted twice. Only the cluster role
differs.

| quantity | patients as units | teams as units |
| --- | --- | --- |
| point estimate | 0.1237 | 0.1256 |
| standard error | 0.0195 | 0.0301 |
| 95% interval | (0.0856, 0.1619) | (0.0667, 0.1845) |

The two point estimates are nearly identical. They differ slightly because the team declaration
also regroups the folds. Step 8 shows how. Declaring the navigator team makes the standard error
1.54 times larger on this draw. Both intervals contain the true ATE of 0.104 on this draw.

The last two lines give the team protocol. Its fingerprint is `7df34d840b4e8d1a`, and its outcome
field names the top-box score. Every other field is the program's.

Neither interval is wrong about the estimand. The unclustered one is wrong about how much
information three thousand patients from 200 teams carry. Cluster-robust inference sums the
influence values inside a team, then takes the variance across teams. With singleton teams it
equals the ordinary formula. [Inference](../technical-reference/inference.md#clusters) gives the
formula.

The large-sample argument counts teams, not patients. The variance includes the factor
$m/(m-1)$ for $m$ teams. With 40 teams or more the interval uses a normal reference.

With 10 to 39 teams the interval uses a $t$ reference with $m-2$ degrees of freedom. Fewer
than 10 teams give a point estimate only. Teams may hold different numbers of patients. A
fold-evaluated fit needs at least 2 teams in every validation fold.
[Clusters](../technical-reference/inference.md#clusters) gives these rules. This fit has 200
teams of 15 patients, so it reports a normal-reference interval.

## Step 8: what the team declaration does to the folds

The cluster role also changes how the rows are split. The code reads the realized split plan of the
team fit and counts how many outer folds each team spans. It then fits four teams with five folds
requested, and it prints the warning and the realized fold count.

In [8]:
import warnings

fold_labels = pd.Series(clustered.split_plan.assignments[0], index=team_frame.index)
folds_per_team = fold_labels.groupby(team_frame["navigator_team"]).nunique()
print("largest number of outer folds one team spans:", folds_per_team.max())

four_teams = team_frame["navigator_team"].drop_duplicates().iloc[:4]
few_teams = team_frame[team_frame["navigator_team"].isin(four_teams)].reset_index(drop=True)
with warnings.catch_warnings(record=True) as caught:
    warnings.simplefilter("always")
    few = (
        CausalStudy(few_teams, design=teams_design, protocol=team_protocol)
        .identify(ATE(reference=0))
        .estimate(method=simple)
    )
for warning in caught:
    print("warning:", warning.message)
print("rows:", few_teams.shape[0], "teams:", few.data.n_clusters)
print("outer folds requested: 5, realized:", few.split_plan.n_folds)

largest number of outer folds one team spans: 1
rows: 60 teams: 4
outer folds requested: 5, realized: 4


**What this output tells you.** Each team spans at most 1 outer fold. With four teams and five
folds requested, the fit warns and realizes 4 folds.

| what happens | why it matters |
| --- | --- |
| each team lands entirely in one outer fold | a patient's nuisance prediction cannot use outcomes from the patient's own team |
| with fewer teams than folds, the fit reduces the fold count and warns | the number of teams, not the number of patients, bounds the split |

The registered
[clustered CV-TMLE study](../technical-reference/method-evidence/clustered-point-treatment-cv-tmle.md#theory-properties)
compares cluster-robust and independent-row variances on the same influence curves. The trust
table below names its design.

## Step 9: reuse the same outer split

Return to the unclustered `cross_fitted` result from Step 5. A second fit can run on exactly its
validation rows. Use this when you compare two methods rather than two splits.

Every point-treatment result exposes its realized outer folds as `split_plan`. Pass that plan back
through `CrossFitting`. The fits below use a different runtime seed, so only the plan can hold the
folds fixed. The last fit compares a second method on the same folds: a main-terms logistic
treatment learner in place of the shallow booster.

In [9]:
new_seed = Runtime(random_state=35, n_jobs=1)
plan = cross_fitted.split_plan
redrawn = effect.estimate(
    method=TMLEMethod(
        models=boosted,
        cross_fitting=CrossFitting(n_folds=5),
        targeting=declared_support,
        runtime=new_seed,
    )
)
reused = effect.estimate(
    method=TMLEMethod(
        models=boosted,
        cross_fitting=CrossFitting(n_folds=5, split_plan=plan),
        targeting=declared_support,
        runtime=new_seed,
    )
)
logistic_g = effect.estimate(
    method=TMLEMethod(
        models=ModelSpec(
            outcome_learner=HistGradientBoostingRegressor(random_state=34),
            treatment_learner=LogisticRegression(max_iter=1000, random_state=34),
        ),
        cross_fitting=CrossFitting(n_folds=5, split_plan=plan),
        targeting=declared_support,
        runtime=new_seed,
    )
)
print(plan)
print("fold fingerprint, Step 5 fit:     ", cross_fitted.provenance.fold_fingerprint)
print("fold fingerprint, new seed:       ", redrawn.provenance.fold_fingerprint)
print("fold fingerprint, new seed + plan:", reused.provenance.fold_fingerprint)
print("fold fingerprint, logistic g:     ", logistic_g.provenance.fold_fingerprint)
print("same point estimate as Step 5:", reused["ate"].psi == cross_fitted["ate"].psi)
print(
    "same influence curve as Step 5:",
    np.array_equal(reused["ate"].influence_curve, cross_fitted["ate"].influence_curve),
)
reuse_comparison = side_by_side(
    {
        "Step 5 fit": cross_fitted,
        "new seed, new folds": redrawn,
        "new seed, same plan": reused,
        "logistic g, same plan": logistic_g,
    }
)
print(reuse_comparison.round(5))

SplitPlan(n=3000, n_folds=5, n_repeats=1, fingerprint=c7f4eafaaf319249, generator=cleverly.random_partition/1, source=32c6f838ef95b45f)
fold fingerprint, Step 5 fit:      c7f4eafaaf319249
fold fingerprint, new seed:        8fc3286eb5d36dd0
fold fingerprint, new seed + plan: c7f4eafaaf319249
fold fingerprint, logistic g:      c7f4eafaaf319249
same point estimate as Step 5: True
same influence curve as Step 5: True
                           psi  std_err  ci_lower  ci_upper
Step 5 fit             0.15731  0.00544   0.14664   0.16798
new seed, new folds    0.15795  0.00544   0.14729   0.16862
new seed, same plan    0.15731  0.00544   0.14664   0.16798
logistic g, same plan  0.15721  0.00520   0.14701   0.16740


**What this output tells you.** Read the output from the top.

| output line | what it shows |
| --- | --- |
| `SplitPlan(...)` | 3000 rows, 5 folds, 1 repeat, the generator `cleverly.random_partition/1`, and `source=32c6f838ef95b45f`, the data fingerprint from Step 5 |
| fold fingerprints | the new seed alone draws new folds. The plan repeats `c7f4eafaaf319249` exactly, for the reused fit and for the logistic fit |
| same point estimate, same influence curve | both `True`. The plan reproduces the Step 5 fit, although the seed changed |
| the table | the new folds move the estimate to 0.15795. The reused plan gives 0.15731 again. The logistic treatment learner gives 0.15721 on the same folds |

The two treatment learners now differ only in the model, not in the split. On this draw their
estimates agree to three decimals. The comparison does not show which treatment model is right.

The `generator` field records how the plan was drawn. `cleverly` draws every unstratified split
itself. It records the generator, the scheme, the requested fold count, and the seed. Take the plan
from a result, as this step does.

A plan belongs to the data it was drawn on. It labels rows by position, and its `source` field
records the data fingerprint. Reuse a plan only on the same rows in the same order. The
[reusable outer split plans](../technical-reference/cv-tmle.md#reusable-outer-split-plans) section
states the contract.

This is reproducible wiring. It adds no evidence about bias, coverage, or efficiency.

## Step 10: other constructions over the same folds

`cleverly` ships three constructions over these folds. They are different estimators, and each has
its own registered evidence.

| construction | how it is selected | what it does |
| --- | --- | --- |
| [stacked CV-TMLE](../technical-reference/method-evidence/stacked-point-treatment-cv-tmle.md) | `targeting_scheme="pooled"`, the default | stacks all out-of-fold predictions, fits one targeting regression, and evaluates the plug-in on the whole sample |
| [fold-evaluated CV-TMLE](../technical-reference/method-evidence/fold-evaluated-point-treatment-cv-tmle.md) | `CrossFitting(fold_evaluation=True)` | keeps the pooled update, averages the fold plug-ins, and uses a cross-validated variance |
| [fold-targeted CV-TMLE](../technical-reference/method-evidence/fold-targeted-point-treatment-cv-tmle.md) | `CrossFitting(targeting_scheme="fold")` | fits one targeting regression inside each validation fold |

The code fits the two alternatives on the Step 5 plan, so all three use identical folds. It prints
the construction line of each summary and the estimates. The
[variations table](../technical-reference/cv-tmle.md#variations) states the scope of each one.

In [10]:
alternatives = {
    "fold-evaluated": CrossFitting(n_folds=5, fold_evaluation=True, split_plan=plan),
    "fold-targeted": CrossFitting(n_folds=5, targeting_scheme="fold", split_plan=plan),
}
construction_fits = {"stacked": cross_fitted}
for label, folds in alternatives.items():
    construction_fits[label] = effect.estimate(
        method=TMLEMethod(
            models=boosted,
            cross_fitting=folds,
            targeting=declared_support,
            runtime=Runtime(random_state=34, n_jobs=1),
        )
    )
for label, fitted in construction_fits.items():
    print(f"{label}: {summary_line(fitted, ': nuisances ')}")
construction_comparison = side_by_side(construction_fits)
print(construction_comparison.round(5))

stacked: stacked CV-TMLE (Levy): nuisances cross-fitted over 5 folds; targeting: pooled
fold-evaluated: fold-evaluated CV-TMLE: nuisances cross-fitted over 5 folds; targeting: pooled
fold-targeted: fold-specific targeted TMLE: nuisances cross-fitted over 5 folds; targeting: fold
                    psi  std_err  ci_lower  ci_upper
stacked         0.15731  0.00544   0.14664   0.16798
fold-evaluated  0.15731  0.00544   0.14664   0.16798
fold-targeted   0.15730  0.00543   0.14666   0.16793


**What this output tells you.** Each summary names its own construction. The fold-targeted fit calls
itself `fold-specific targeted TMLE`, with `targeting: fold`.

| construction | estimate | standard error |
| --- | --- | --- |
| stacked | 0.15731 | 0.00544 |
| fold-evaluated | 0.15731 | 0.00544 |
| fold-targeted | 0.15730 | 0.00543 |

The fold-evaluated fit keeps the stacked update and averages the plug-ins of equal-size folds. Its
point estimate therefore equals the stacked one. Only its variance formula differs, and on this
draw the two intervals agree to five decimals. The fold-targeted fit refits the targeting inside
each validation fold. On this draw, its estimate and its standard error differ from the stacked
ones in the fifth decimal.

One draw does not rank the three estimators. Each construction's evidence page reports its own
repeated-sampling results.

## Step 11: diagnostics, what the fit can check

Start with the combined assessment of the Step 5 fit. It presents validation, diagnostics, and
sensitivity together. `assessment.attention` lists the rows that need review. The code then prints
three retained reports: support, nuisance models, and score equations. The last line gives the
support detail of the in-sample fit for comparison.

In [11]:
assessment = cross_fitted.assess()
print(assessment.summary())
print("needs attention:", tuple(item.name for item in assessment.attention))
support = assessment.report("support")
nuisance = assessment.report("nuisance_models")
scores = assessment.report("score_equations")
print()
print(support.summary())
print()
print(nuisance.summary())
print()
print(scores.summary())
print()
smallest = support.propensity_quantiles["overall"][0.0]
print(f"smallest estimated g(W), before truncation: {smallest:.4f}")
truncated = support.truncated
print(f"truncated: {truncated['count']:.0f} units ({truncated['fraction']:.2%})")
for arm, ess in support.effective_sample_size.items():
    ratio = f"{ess['effective']:.1f} ({ess['ratio']:.1%})"
    print(f"{arm}: {ess['n']:.0f} rows, effective sample size {ratio}")
print()
print("in-sample support:", in_sample.diagnostics.run_all()["support"].detail)

Returned results
----------------
surface      operation            result                                                                                                                                                                                    
-----------  -------------------  ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
validation   support              maximum truncated fraction 0.0%; minimum effective-sample-size ratio 82.5%; group load: mean:h1 1137.7/3000 Kish-equivalent mask rows (37.9%; 37.9% all; draw 01 of 01); not estimator ESS
validation   nuisance_models      2 nuisance model report(s) are available                                                                                                                                                  
sensitivity  omitted_confounding  at the default strengths; cf_y=0.03, cf_d=0.03, 

**What this output tells you.** Read the parts in order.

| output part | what it shows on this draw |
| --- | --- |
| `Checks` and `needs attention` | no row needs attention. The score-equation check passed |
| nuisance model diagnostics | the propensity's out-of-fold calibration slope is 0.9611. The verdict reads `nuisance fits look reasonable` |
| support fields | the smallest fitted g(W) is 0.0239, and no unit sits at the truncation bound. The treated arm keeps an effective-sample-size ratio of 82.5% |
| in-sample support | the in-sample fit reports a minimum effective-sample-size ratio of 89.6% |

The `Returned results` rows are calculations that ran. They are not passes. Their sensitivity rows
use the default strengths, and Step 13 reads the same analysis. The nuisance report measures the
models on patients they did not see. No row needed attention on 199 of the 200 probe draws. The
calibration slope lay between 0.90 and 1.01 on 90% of them.

These reports describe the fitted propensity model, not the treatment law. A calibration slope
near 1 does not show that the model is correct. The fitted model puts a share of 0.0063 of the rows
below 0.1, in the `P(g < t)` column. In the law, about 1.9% of patients have a true propensity
below 0.1 (review probe `reviews/notebook-review/probes/cross-fitting-final/truth.py`). The fitted model does not see that tail. Its share
fell below the law's share on 196 of the 200 probe draws.

`cleverly` derives no threshold for the effective sample size, so the support report gives 82.5% no
grade.

The in-sample ratio is higher, although the true assignment mechanism is the same. Its
observed-arm weights are closer to one. The in-sample ratio exceeded the cross-fitted one on all 200 probe draws.
Cross-fitting does not repair poor support. The support report describes fitted overlap and cannot
verify population positivity.

Positivity means that every kind of patient has some chance of each arm. The
[diagnostics guide](../user-guide/results-assessment.md#diagnostics) describes each report.

## Step 12: the fold draw itself

One split is one draw. `repeats=` runs the complete estimator on several fold draws and reports the
median. This fit reports one estimate, so the call keeps the default `Inference()`.

In [12]:
repeated = effect.estimate(
    method=TMLEMethod(
        models=boosted,
        cross_fitting=CrossFitting(n_folds=5, repeats=3),
        targeting=declared_support,
        runtime=Runtime(random_state=34, n_jobs=1),
    )
)
print(summary_line(repeated, "median over"))
print(side_by_side({"median of 3 draws": repeated}).round(4))
repeated_nuisance = repeated.diagnostics.run_all().report("nuisance_models")
print(repeated_nuisance.repeat_spread_frame().round(4).to_string(index=False))

(median over 3 independent draws of the split; variance includes within-draw uncertainty and split dispersion)
                      psi  std_err  ci_lower  ci_upper
median of 3 draws  0.1567   0.0055    0.1459    0.1674
estimand  n_repeats  standard_deviation  reported_standard_error  ratio_to_standard_error
     ate          3              0.0004                   0.0055                   0.0679


**What this output tells you.** The first line confirms the median over 3 independent draws. The
median estimate is 0.1567 with a standard error of 0.0055.

`repeats=` is the same estimator over several draws, not a new estimator. It reports the median
point, and its variance includes the split dispersion. The
[variations table](../technical-reference/cv-tmle.md#variations) states the scope of a repeated
fit.

The median rule takes the median over the draws of each variance plus the squared distance from
the median point. The reported standard error is 0.0055, against 0.00544 for the single Step 5
draw.

The spread row gives a standard deviation of 0.0004 across the three draws. The reported standard
error is 0.0055, and their ratio is 0.0679. Across these three draws, the estimate moved much less
than its standard error. Three draws give an imprecise spread. In Step 9, one new draw moved the
estimate from 0.15731 to 0.15795. The report defines no threshold for the ratio.

The [repeated CV-TMLE evidence](../technical-reference/method-evidence/repeated-cross-fitting.md)
gives the repeated-sampling results for the median report.

## Step 13: sensitivity, what folds cannot check

Sensitivity analysis asks how strong an unmeasured confounder would need to be to change the
conclusion. Cross-fitting changes the estimator, but it does not change the
no-unmeasured-confounding assumption.

| term | plain meaning |
| --- | --- |
| `cf_y` | the share of the remaining outcome variation that a hidden confounder explains |
| `cf_d` | the share of the Riesz representer's second moment that only the hidden confounder adds |
| `rho` | how closely the confounder's two effects align. `rho=1` is the worst case |
| robustness value | the equal `cf_y` and `cf_d` that moves the point estimate to zero at `rho=1` |
| robustness value at the 95% limit | the equal `cf_y` and `cf_d` that moves the one-sided 95% limit to zero at `rho=1` |
| $\nu^2$ | the second moment of the Riesz representer. For the ATE it is $E[1 / \{g(1 - g)\}]$, the mean inverse variance of the treatment given the covariates |
| bias scale | $\sqrt{\sigma^2 \nu^2}$, where $\sigma^2$ is the residual outcome variance. The bound multiplies it by a factor set by `cf_y`, `cf_d`, and `rho`, and that factor can exceed 1 |

Each bound needs an estimate of $\nu^2$. The code uses the default estimator, which resolves to
the doubly robust form. $\nu^2$ is finite only when $1/g$ and $1/(1 - g)$ have finite means. The
true propensity of this law lies between 5% and 95%, so its $\nu^2$ is finite. It is about 4.83
(review probe `reviews/notebook-review/probes/cross-fitting-final/truth.py`).

The bounds call keeps the package's default strengths, `cf_y = cf_d = 0.03` at `rho = 1`. Those
strengths are not calibrated to this program. The
[point-treatment tutorial](point-treatment-tmle.ipynb#step-10-sensitivity-what-the-fit-cannot-check)
calibrates them against a recorded covariate. The
[sensitivity guide](../user-guide/results-assessment.md#sensitivity-analysis) defines these
quantities.

In [13]:
sensitivity = cross_fitted.sensitivity
elements = sensitivity.elements(estimand="ate")
robustness = sensitivity.robustness_value()
bounds = sensitivity.omitted_confounding()
print(f"sigma^2 (residual outcome variance): {elements.sigma2:.6f}")
print(f"nu^2 (doubly robust):                {elements.nu2:.4f}")
print(f"robustness value:                    {robustness['rv']:.3f}")
print(f"robustness value at the 95% limit:   {robustness['rva']:.3f}")
print()
print(bounds.summary())

sigma^2 (residual outcome variance): 0.018093
nu^2 (doubly robust):                4.5363
robustness value:                    0.419
robustness value at the 95% limit:   0.398

Omitted-variable sensitivity for 'ate'
--------------------------------------------
estimate 0.15731; bias scale sqrt(sigma^2 nu^2) = 0.28648
assumed confounding: cf_y = 0.03, cf_d = 0.03, rho = 1 -> bias <= 0.0087264
bias-adjusted bounds:  [0.14859, 0.16604]
with 95% one-sided CIs: [0.13962, 0.17499] (the sign of the effect survives)

robustness value RV   = 0.4187: a confounder with cf_y = cf_d = 41.9% (the share of the residual outcome variation it explains, and the share of the Riesz representer's second moment only it adds) would move the estimate to 0.
robustness value RVa  = 0.3980: the same, for the 95% confidence bound.


**What this output tells you.** The robustness value is 0.419. At `rho=1`, a confounder with
`cf_y` and `cf_d` both equal to 0.419 would move the point estimate to zero. The robustness value
at the 95% limit is 0.398. Equal strengths of 0.398 would move the one-sided 95% limit to zero.

The bound block reads the default strengths. The bias scale is 0.28648, and the strengths limit
the bias to 0.0087264. The bias-adjusted bounds are [0.14859, 0.16604]. Their one-sided 95% limits
are [0.13962, 0.17499], and the output reads `the sign of the effect survives`. On the 200 probe
draws, these limits excluded zero every time.

The fit printed the doubly robust $\nu^2$ as 4.5363, below the law's 4.83. The estimator causes
this direction. Its expected value is the law's $\nu^2$ minus the squared error of the fitted Riesz
representer. The estimate therefore reads low whenever the fitted propensity has an error. Step 11
shows that the fitted propensities miss part of the law's tail.

On the 200 probe draws, the estimate had a mean of 4.50, about 7% below 4.83. It fell below 4.83 on
195 draws. A $\nu^2$ that is 7% low makes the bias scale about 3% low. The bounds are then slightly
narrow, and the robustness values read slightly high.

On this draw, the law's $\nu^2$ gives a robustness value of 0.409 rather than 0.419. It widens the
bias-adjusted bounds from [0.14859, 0.16604] to [0.1483, 0.1663]. The review probe
`reviews/notebook-review/probes/cross-fitting-final/summarize.py` computes these values.

These numbers are conditional on the omitted-variable model and on its worst-case alignment.
Cross-fitting does not make sensitivity analysis unnecessary. It answers an identification
question that the fold construction cannot check. The
[omitted-variable bounds](../technical-reference/validation-methods.md#omitted-variable-bounds-robustness-value-benchmark-and-contours)
section defines each quantity.

## How far to trust this

| reviewer question | answer on this page |
| --- | --- |
| why split the sample | on this page's law, the registered in-sample control covered 0.5225 of the time and the cross-fitted fit 0.945. On this page's configuration, the review probe found 153 and 188 covering intervals in 200 draws (Step 6). On the shown draw, the in-sample standard error was 0.73 times the cross-fitted one |
| are patients independent | no. The team law gives each team a shared effect. Declaring teams made the standard error 1.54 times larger on this draw and kept each team in one fold (Steps 7 and 8) |

Two registered studies cover the constructions of this page. The
[stacked point-treatment CV-TMLE study](../technical-reference/method-evidence/stacked-point-treatment-cv-tmle.md)
runs its `crossfit_overfitting` cells on this page's law at n = 500. Each cell has 400 replicates,
ten folds, and propensity bounds of 0.025 and 0.975. The
[clustered CV-TMLE study](../technical-reference/method-evidence/clustered-point-treatment-cv-tmle.md)
runs its `clustered_inference` cells on the team law with teams of 10, at n = 2000, with 2400
replicates and five grouped folds.

| cited cell | law | nuisance construction | outcome type | fold layer | role | coverage of the 95% interval |
| --- | --- | --- | --- | --- | --- | --- |
| `crossfit_overfitting/stacked_cvtmle` | this page's Step 2 law | fitted: an unpruned regression tree for Q and a main-terms logistic g | bounded continuous | cross-fitted | positive | 0.945 |
| `crossfit_overfitting/in_sample_control` | this page's Step 2 law | fitted: the same two learners | bounded continuous | in sample | control | 0.5225 |
| `clustered_inference/cluster_robust` | the Step 7 team law, with teams of 10 rather than 15 | oracle g, and a fitted main-terms logistic Q | binary | cross-fitted | positive | 0.9475 |
| `clustered_inference/iid_control` | the same law | the same fits, with an independent-row variance | binary | cross-fitted | control | 0.84125 |

The coverage values come from each study's `properties.csv`. The in-sample tree sees the rows it
predicts, and its interval covers about half the time. Cross-fitting the same tree restores
coverage near 95%. On the team law, the independent-row variance undercovers, and the
cluster-robust variance covers near 95%.

No registered study covers this page's boosted learners or five folds on the Step 2 law. The review
probe `reviews/notebook-review/probes/cross-fitting-final/sweep.py` covers that configuration, and it is not a registered study. The
clustered study supplies the exact treatment mechanism, while Step 7 fits it.

| layer | establishes | does not establish |
| --- | --- | --- |
| the two comparisons in Steps 6 and 7 | that in-sample nuisances and undeclared teams each gave a narrower interval on this draw | the coverage rate of any of these estimators |
| the out-of-fold nuisance report | how the learners performed on unseen patients | that the learners converge fast enough for the remainder condition |
| the retained support report | truncation and effective sample size of the fitted propensity model | that the fitted model sees the law's tail, or that cross-fitting repairs poor support |
| the retained split-spread rows | how much the estimate moved across the three declared fold draws | whether that amount is acceptable or another draw would help |
| the omitted-confounding analysis | the bound under its declared strength and alignment, if the fitted propensity is right | that no stronger hidden confounder exists, or that the fitted $\nu^2$ equals the law's |
| the registered studies | each construction passes its accuracy and property rows, within the limits its page declares | that your identification assumptions hold on your data, or coverage for learners and fold counts no row covers |

Nothing in this list validates the causal reading. That rests on consistency, no interference, and
no unmeasured confounding. The
[technical entry](../technical-reference/cv-tmle.md#validation-issues-special-to-this-method)
links the evidence for each construction.

## Where to go next

Cross-fitting addresses only data reuse. If you doubt one nuisance and still want an interval, the
condition you are worried about is the
[product rate](../technical-reference/cv-tmle.md#what-this-solves). The variant for it is
[DR-TMLE](dr-tmle.ipynb). If your worry is instead which baseline variables belong in the assignment
model, read [collaborative TMLE](collaborative-tmle.ipynb).

The [examples index](index.md#the-program) lists every tutorial in the program.